File: 02_first_node/first_node.cpp
#include "rclcpp/rclcpp.hpp"

int main(int argc, char * argv[])
{
    // 1. 初始化 ROS 2 C++ 客户端库
    rclcpp::init(argc, argv);

    // 2. 创建节点
    auto node = std::make_shared<rclcpp::Node>("first_node_cpp");

    // 3. 输出日志
    RCLCPP_INFO(node->get_logger(), "C++实现的第一个自定义节点~");

    // 4. 保持节点运行，等待并处理 ROS 2 事件
    rclcpp::spin(node);

    // 5. 关闭 ROS 2 C++ 客户端库
    rclcpp::shutdown();

    return 0;
}

File: 02_first_node/first_node.py
import rclpy
from rclpy.node import Node

def main(args=None):
    # 1. 初始化 ROS 2 Python 客户端库
    rclpy.init(args=args)

    # 2. 创建节点
    node = Node('first_node_python')

    # 3. 输出日志
    node.get_logger().info('Python实现的第一个自定义节点~')

    # 4. 保持节点运行，等待并处理 ROS 2 事件
    rclpy.spin(node)

    # 5. 关闭 ROS 2 Python 客户端库
    rclpy.shutdown()

if __name__ == '__main__':
    main()
    